### PR TITLE
[MIG] stock_by_warehouse: Migration to 16.0 T#63809

### PR DESCRIPTION
**Pull request by rolandojduartem**
Minor changes are applied to the code to 16.0, but also, the widget
'warehouse' is migrated to OWL.

**Before recompute:**

Screenshot from 2023-03-17 16-05-16

**After recompute:**

Screenshot from 2023-03-17 16-02-32

Screenshot from 2023-03-17 16-02-40

### CODE REVIEW

**Comment by rolandojduartem** [comment]
Hi, @luisg123v and @imanie383, could you review this migration? the widget 'warehouse' was migrated in this module. The instance for the functional review will be deployed.

cc @desdelinux

**Comment by rolandojduartem** [comment]
@desdelinux, for the functional review, you could use the runbot on https://git.vauxoo.com/vauxoo/dhm/-/merge_requests/130 or https://git.vauxoo.com/vauxoo/dhm/-/merge_requests/131

File: stock_by_warehouse/static/src/components/warehouse/template.xml [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+<templates xml:space="preserve">
+
+    <t t-name="stock_by_warhouse.ShowWarehouseInfo" owl="1">
+        <div>
+            <a role="button" tabindex="0" class="js_product_warehouse" t-on-click.stop="(ev) => this.onClick(ev)">
+                <b t-out="show" /> ... see all warehouses .. <i class="fa fa-list-ul" /></a>
+        </div>
+    </t>
+
+    <t t-name="stock_by_warhouse.ProductWarehousePopOver" owl="1">
```
Comment (luisg123v) [inline stock_by_warehouse/static/src/components/warehouse/template.xml]:
```suggestion
    <t t-name="stock_by_warehouse.ProductWarehousePopOver" owl="1">
```

Same in other appearances

File: stock_by_warehouse/static/src/components/warehouse/widget.js [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+    return Object.assign(textExtractProps({attrs, field}), {
+        byLocation: archParseBoolean(attrs.options.by_location),
+    });
+};
+
+// Pop-up to show the information detailed by the warehouses
+export class ProductWarehousePopOver extends Component {}
+ProductWarehousePopOver.template = "stock_by_warhouse.ProductWarehousePopOver";
+
+// Pop-up to show the information detailed by the lacations
```
Comment (luisg123v) [inline stock_by_warehouse/static/src/components/warehouse/widget.js]: locations

File: stock_by_warehouse/static/src/components/warehouse/widget.js [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+        if (!info) {
+            return;
+        }
+        this.info = info;
+        this.show = formatFloat(this.props.byLocation ? this.info.available_locations : this.info.warehouse, {
+            digit: 2,
+        });
+        this.lines = this.info.content || [];
+        for (const value of this.lines) {
+            value.available_not_res_formated = formatFloat(value.available_not_res, {digits: 2});
```
Comment (luisg123v) [inline stock_by_warehouse/static/src/components/warehouse/widget.js]: format**t**ed.

Same in all other appearances

File: stock_by_warehouse/static/src/components/warehouse/template.xml [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+        <h3 t-if="this.props.title" class="o_popover_header" t-out="this.props.title" />
+        <div class="popover-body stock-by-warehouse-widget">
+            <div class="panel clearfix container">
+                <div class="list-group">
+                    <t t-foreach="this.props.lines" t-as="line" t-key="line_index">
+                        <div
+                            t-if="line.available_not_res &gt; 0 or line.available &gt; 0 or line.outgoing &gt; 0 or line.saleable &gt; 0 or line.incoming &gt; 0 or line.virtual &gt; 0"
```
Comment (luisg123v) [inline stock_by_warehouse/static/src/components/warehouse/template.xml]: Please split

File: stock_by_warehouse/static/src/components/warehouse/template.xml [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+                                </b>
+                                <div class="text-right">
+                                    <span
+                                        title="This is real(on hand) quantity  minus reserved quantity"
+                                        class="text-muted"
+                                        t-out="line.available_not_res_formated"
+                                    />
+                                    <br />
+                                    <span
+                                        title="Available Now but not necessarily what I can sale"
```
Comment (luisg123v) [inline stock_by_warehouse/static/src/components/warehouse/template.xml]: Update translations

File: stock_by_warehouse/static/src/components/warehouse/widget.js [hunk 6]
```diff
@@ -0,0 +1,91 @@
...
+// Pop-up to show the information detailed by the warehouses
+export class ProductWarehousePopOver extends Component {}
+ProductWarehousePopOver.template = "stock_by_warhouse.ProductWarehousePopOver";
+
+// Pop-up to show the information detailed by the lacations
+export class StockAvailabilityPopOver extends Component {}
+StockAvailabilityPopOver.template = "stock_by_warhouse.StockAvailabilityPopOver";
+
+// Main Widget
+export class StockByWarehouse extends TextField {
```
Comment (luisg123v) [inline stock_by_warehouse/static/src/components/warehouse/widget.js]:
```suggestion
export class StockByWarehouseField extends TextField {
```

File: stock_by_warehouse/static/src/components/warehouse/widget.js [hunk 7]
```diff
@@ -0,0 +1,91 @@
...
+import {usePopover} from "@web/core/popover/popover_hook";
+import {localization} from "@web/core/l10n/localization";
+import {formatFloat} from "@web/views/fields/formatters";
+import {TextField} from "@web/views/fields/text/text_field";
+import {archParseBoolean} from "@web/views/utils";
+
+const {Component, onWillUpdateProps} = owl;
+
+// Add the new option by_location to the props of TextField
+TextField.props = {
```
Comment (luisg123v) [inline stock_by_warehouse/static/src/components/warehouse/widget.js]: I think you should modify the new component, not the native one

File: stock_by_warehouse/__manifest__.py [hunk 8]
```diff
@@ -0,0 +1,29 @@
...
+    ],
+    "data": [
+        "views/product_product_views.xml",
+        "views/product_template_views.xml",
+    ],
+    "demo": [],
+    "assets": {
+        "web.assets_backend": [
+            "stock_by_warehouse/static/src/components/warehouse/main.scss",
+            "stock_by_warehouse/static/src/components/warehouse/template.xml",
```
Comment (luisg123v) [inline stock_by_warehouse/__manifest__.py]:
```suggestion
            "stock_by_warehouse/static/src/components/warehouse/warehouse_field.xml",
```

Same for the other files inside the same directory

**Review by luisg123v — CHANGES_REQUESTED** [review]

File: stock_by_warehouse/i18n/es.po [hunk 9]
```diff
@@ -0,0 +1,173 @@
...
+#. odoo-javascript
+#: code:addons/stock_by_warehouse/static/src/components/warehouse/warehouse_field.xml:0
+#, python-format
+msgid "Outgoing"
+msgstr "Saliente"
+
+#. module: stock_by_warehouse
+#: model:ir.model,name:stock_by_warehouse.model_product_template
+msgid "Product"
+msgstr "Plantilla de producto"
```
Comment (luisg123v) [inline stock_by_warehouse/i18n/es.po]: It doesn't match.

The same occurs in product variants

File: stock_by_warehouse/static/src/components/warehouse/warehouse_field.js [hunk 10]
```diff
@@ -0,0 +1,100 @@
...
+// Main Widget
+export class StockByWarehouseField extends TextField {
+    setup() {
+        this.initalizeVariables();
+        this.popover = usePopover();
+        this.formatData(this.props);
+        onWillUpdateProps((nextProps) => this.formatData(nextProps));
+    }
+
+    initalizeVariables() {
```
Comment (luisg123v) [inline stock_by_warehouse/static/src/components/warehouse/warehouse_field.js]:
```suggestion
    initializeVariables() {
```

File: stock_by_warehouse/static/src/components/warehouse/warehouse_field.xml [hunk 11]
```diff
@@ -0,0 +1,117 @@
...
+    <t t-name="stock_by_warehouse.ProductWarehousePopOver" owl="1">
+        <h3 t-if="this.props.title" class="o_popover_header" t-out="this.props.title" />
+        <div class="popover-body stock-by-warehouse-widget">
+            <div class="panel clearfix container">
+                <div class="list-group">
+                    <t t-foreach="this.props.lines" t-as="line" t-key="line_index">
+                        <div
+                            t-if="
+                                line.available_not_res
+                                &gt; 0 or line.available
```
Comment (luisg123v) [inline stock_by_warehouse/static/src/components/warehouse/warehouse_field.xml]: One condition per line, please, currently conditions are split, half on on a line and the other alf in the next one

**Review by luisg123v — CHANGES_REQUESTED** [review]

**Review by desdelinux — APPROVED** [review]

**Comment by desdelinux** [comment]
Functional review is ok.

cc @hugho-ad @rolandojduartem @luisg123v

**Review by hugho-ad — APPROVED** [review]

**Comment by hugho-ad** [comment]
LGTM

**Comment by rolandojduartem** [comment]
Hi, @luisg123v, the changes were applied to the code, please, could you review again?

File: stock_by_warehouse/i18n/es.po (hunk 9 above)
Comment (luisg123v) [reply stock_by_warehouse/i18n/es.po]: I've double checked this and, in this case, what doesn't match is the original term. Check the `product` module

Comment (luisg123v) [reply stock_by_warehouse/i18n/es.po]: Oh, I see, original term has changed, see https://github.com/odoo/odoo/commit/c2f394253ef329c62497b8d9986b36183354e474#diff-921ec733c3821301f62cd0d96e45990e7636f5db56a10ae76aed009e380bf149R18

 The one that it's outdated (this comes from Odoo) is the translation.

Let's fix the translation here and in Odoo.

Comment (rolandojduartem) [reply stock_by_warehouse/i18n/es.po]: Hi, @luisg123v , ready and the link I think is https://www.transifex.com/odoo/odoo-16/translate/#es/product/437962391

**Comment by rolandojduartem** [comment]
Hi, @luisg123v, please, could you review this PR?

**Review by luisg123v — APPROVED** [review]

**Pull request merged**